tuf/repository_lib.py

```diff
@@ -198,13 +206,28 @@ def _generate_and_write_metadata(rolename, metadata_filename, write_partial,
       roleinfo = tuf.roledb.get_roleinfo(rolename)
       roleinfo['version'] = roleinfo['version'] + 1
       tuf.roledb.update_roleinfo(rolename, roleinfo) 
-      signable = sign_metadata(metadata, roleinfo['signing_keyids'],
-                               metadata_filename)
+      signable = sign_metadata(metadata, signing_keyids, metadata_filename)
   
   # Write the metadata to file if it contains a threshold of signatures. 
   signable['signatures'].extend(roleinfo['signatures']) 
+
   
-  if tuf.sig.verify(signable, rolename) or write_partial:
+  def should_write():
+    # Always write if partial writing
+    if write_partial:
+      return True
+    
+    # In the normal case, we should write metadata if the threshold is meet
+    write = tuf.sig.verify_threshold(signable, roleinfo['threshold'], roleinfo['signing_keyids'], rolename)
+    
+    # The root must also be signed by previous root keys and threshold
+    if rolename == 'root' and len(previous_keyids) > 0 and previous_threshold is not None:
```

Comment:
I think `previous_threshold` might never be set `None`.

According to the following line (on line 125), it can either be 1 or a valid threshold number enforced by tuf.roledb:
```previous_threshold = roleinfo.get('previous_threshold', 1)```